core/arch/arm/kernel/spmc_sp_handler.c

```diff
@@ -216,10 +216,87 @@ static int spmc_sp_add_sp_region(struct sp_mem *smem,
 	return res;
 }
 
+static int release_mobj(struct sp_mem *smem, struct sp_mem_map_region *region,
+			bool mobj_ffa)
+{
+	uint64_t handle = smem->transaction.global_handle;
+	int res = FFA_OK;
+
+	if (mobj_ffa) {
+		if (handle) {
+			mobj_put(region->mobj);
+			if (mobj_ffa_unregister_by_cookie(handle))
+				return FFA_DENIED;
+
+			if (mobj_ffa_sel1_spmc_reclaim(handle))
+				return FFA_DENIED;
+		} else {
+			mobj_ffa_sel1_spmc_delete(to_mobj_ffa(region->mobj));
+		}
+	} else {
+		mobj_put(region->mobj);
+	}
+
+	return res;
+}
+
 static int spmc_sp_add_nw_region(struct sp_mem *smem,
 				 struct ffa_mem_region *mem_reg)
 {
+	uint64_t page_count = READ_ONCE(mem_reg->total_page_count);
+	struct mobj *mobj = NULL;
+	struct sp_mem_map_region *region = NULL;
+	struct mobj_ffa *m = mobj_ffa_sel1_spmc_new(page_count);
```

Comment:
`mobj_ffa_sel1_spmc_new()` designed to handle FFA_MEM_SHARE with OP-TEE Core. This allows mapping shared memory as found by cookie as argument for any pseudo or user TA. There's also life cycle management for these objects tailored for this usage.

I don't doubt that some parts of mobj_ffa.c can be re-used if desired, but there's also parts which should not be mixed up with these SPs. For example two lists `shm_head` and `shm_inactive_head` should probably not be used for SPs.

Reply:
Changed it so we use our own mobj for this